chat_shell/chat_shell/tools/builtin/knowledge_listing.py

```
@@ -610,3 +653,119 @@ async def _read_docs_http_mode(
+
+        Supports both package mode (direct DB access) and HTTP mode (via API).
+
+        Args:
+            kb_id: Knowledge base ID
+            document_ids: List of document IDs that were read
+            total_chars_read: Total characters read in this call
+        """
+        # Try package mode first (direct DB access)
+        try:
+            await self._persist_kb_head_package_mode(kb_id, document_ids, total_chars_read)
+            return
```

Ruff (0.14.14)

[warning] Consider moving this statement to an `else` block

(TRY300)